lgc/util/ElfNoteEntryInsertionUtil.cpp

```diff
@@ -0,0 +1,258 @@
+//                            All sections after this offset will be shifted.
+// @param lengthToBeShifted   : The length how much sections after shiftStartingOffset will be shift.
+// @param sectionHeaders      : The array of section headers.
+// @returns                   : The array of the contents of sections to be shifted and their new offset
+//                            sorted by the new offset in the increasing order.
+SmallVector<SectionShiftInfo> updateSectionOffsetsForShift(const SmallVectorImpl<char> &elf,
+                                                           const ELF::Elf64_Off shiftStartingOffset,
+                                                           ELF::Elf64_Off lengthToBeShifted,
+                                                           MutableArrayRef<ELF::Elf64_Shdr> sectionHeaders) {
+  SmallVector<SectionShiftInfo> sectionAndNewOffset;
+
+  // If a section is located after shiftStartingOffset, it must be shifted.
+  for (auto &sectionHeader : sectionHeaders) {
+    if (sectionHeader.sh_offset < shiftStartingOffset)
+      continue;
+    const auto newOffset = alignTo(sectionHeader.sh_offset + lengthToBeShifted, Align(sectionHeader.sh_addralign));
+    sectionAndNewOffset.push_back(
+        {SmallString<64>(StringRef(elf.data() + sectionHeader.sh_offset, sectionHeader.sh_size)), newOffset});
+    lengthToBeShifted = newOffset - sectionHeader.sh_offset;
+
+    // Update the offset of section pointed by the section header to its new offset.
+    sectionHeader.sh_offset = newOffset;
+  }
+
+  // Sort sectionAndNewOffset by the new offset of each section in the increasing order.
+  sort(sectionAndNewOffset,
+       [](const SectionShiftInfo &i0, const SectionShiftInfo &i1) { return i0.newOffset < i1.newOffset; });
+  return sectionAndNewOffset;
+}
+
+// =====================================================================================================================
+// Inserts the new contents to the given ELF.
+//
+// @param [in/out] elf        : The ELF to insert the new contents.
+// @param insertionOffset     : The offset of the new contents to be inserted.
+// @param elfContentStream    : The ELF contents to be inserted.
+// @param sectionAndNewOffset : The sections to be shifted and their new offsets.
+void insertContentsToELF(SmallVectorImpl<char> &elf, const ELF::Elf64_Off insertionOffset,
+                         AppendingBinaryByteStream &elfContentStream,
+                         const SmallVectorImpl<SectionShiftInfo> &sectionAndNewOffset) {
+  // Strip sections after the insertion offset of the new contents.
+  elf.resize(insertionOffset);
+
+  // Write the new contents.
+  raw_svector_ostream elfStream(elf);
+  ArrayRef<uint8_t> noteEntry;
+  auto error = elfContentStream.readBytes(0, elfContentStream.getLength(), noteEntry);
+  (void)error;
+  assert(!error);
+  elfStream << StringRef(reinterpret_cast<const char *>(noteEntry.data()), noteEntry.size());
```

Comment:
You can use `toStringRef(noteEntry)` from `StringExtras.h`. Similarly, there's also `arrayRefFromStringRef` to go the other way.